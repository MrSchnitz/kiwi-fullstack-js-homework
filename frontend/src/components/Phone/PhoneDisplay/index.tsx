import React from "react";
import { Textarea } from "@kiwicom/orbit-components/lib";
import styled, {keyframes} from "styled-components";

interface PhoneDisplayProps {
  text: string;
  onChange?: (value: string) => void;
}

const PhoneDisplay: React.FC<PhoneDisplayProps> = ({
  text,
  onChange,
}: PhoneDisplayProps) => {
  return (
    <TextareaLayout>
      <Textarea
        size={"normal"}
        readOnly={true}
        fullHeight={true}
        placeholder="Type something..."
        value={text.length > 0 ? text : "Type something..."}
      />
    </TextareaLayout>
  );
};

const TextareaLayoutAnimation = keyframes`
    0% {
        opacity: 0;
    }
    100% {
        opacity: 1;
    }
`;

const TextareaLayout = styled.div`
  height: 100%;
  width: 100%;
  font-size: 3rem;
  
  animation: ${TextareaLayoutAnimation} 4s ease;

  textarea {
    font-size: 2.5rem;
    padding: 1.5rem;
    border: none;
    outline: none;
    border-color: transparent;
    background-color: #d6ead9;
    // background-color: #ECF8F7;
  }
`;

export default PhoneDisplay;
