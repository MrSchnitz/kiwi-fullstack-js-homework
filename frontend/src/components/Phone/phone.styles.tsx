import styled, { keyframes } from "styled-components";

export const PhoneContent = styled.div`
  height: 100%;
  width: 100%;
  overflow: hidden;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
`;

PhoneContent.displayName = "PhoneContent";

export const PhoneLockScreenAnimation = keyframes`
    0% {
        transform: scale(1);
    }
    100% {
        transform: scale(1.2);
    }
`;

export const PhoneLockScreenPulseAnimation = keyframes`
    0% {
        opacity: 1;
        transform: scale(.33);
    }
    80%, 100% {
        opacity: 0;
    }
`;

export const PhoneLockedScreen = styled.div`
  height: 100%;
  width: 100%;
  display: flex;
  justify-content: center;
  align-items: center;
  background-color: #000;

  cursor: pointer;

  h1 {
    position: relative;
    color: #fff;

    &:after {
      content: "";
      position: absolute;
      left: -5rem;
      top: -11rem;
      width: 25rem;
      height: 25rem;
      border-radius: 12.5rem;
      border: 3px solid #00a991;
      opacity: 0;

      animation: ${PhoneLockScreenPulseAnimation} 3s
        cubic-bezier(0.215, 0.61, 0.355, 1) infinite;
    }
    animation: ${PhoneLockScreenAnimation} 1.5s alternate infinite;
  }
`;

PhoneLockedScreen.displayName = "PhoneLockedScreen";
