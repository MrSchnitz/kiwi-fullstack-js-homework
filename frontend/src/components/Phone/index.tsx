import React from "react";
import styled from "styled-components";
import Keyboard from "../Keyboard";
import PhoneDisplay from "./PhoneDisplay";
import PhoneLayout from "./PhoneLayout";

interface PhoneProps {
  displayText: string;
  onKeyPressed: (key: number) => void;
  onDeleteChar: () => void;
}

const Phone: React.FC<PhoneProps> = ({
  displayText,
  onKeyPressed,
  onDeleteChar,
}: PhoneProps) => {

  return (
    <PhoneLayout>
      <PhoneContent>
        <PhoneDisplay text={displayText} />
        <Keyboard
          onKeyPressed={onKeyPressed}
          onDeleteChar={onDeleteChar}
          show={true}
        />
      </PhoneContent>
    </PhoneLayout>
  );
};

const PhoneContent = styled.div`
  height: 100%;
  width: 100%;
  overflow: hidden;
  display: flex;
  flex-direction: column;
  // justify-content: center;
  align-items: center;
`;

export default Phone;
