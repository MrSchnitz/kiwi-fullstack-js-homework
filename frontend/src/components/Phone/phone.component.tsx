import React, {useState} from "react";
import PhoneLayout from "../PhoneLayout/phone-layout.component";
import PhoneDisplay from "../PhoneDisplay/phone-display.component";
import Keyboard from "../Keyboard/keyboard.component";
import PredictionSection from "../PredictionSection/prediction-section.component";
import {WordType} from "../../models/Word";
import {PhoneContent, PhoneLockedScreen} from "./phone.styles";

interface PhoneProps {
  displayText: string;
  predictionWords: WordType[];
  predictionWordsLoading?: boolean;
  onKeyPressed: (key: number) => void;
  onDeleteChar: () => void;
  onPredictionSelect: (word: WordType) => void;
}

const Phone: React.FC<PhoneProps> = ({
  displayText,
  predictionWords,
  predictionWordsLoading,
  onKeyPressed,
  onDeleteChar,
  onPredictionSelect,
}: PhoneProps) => {
  const [isUnLocked, setIsUnLocked] = useState(true);

  const renderLockedScreen = (
    <PhoneLockedScreen onClick={() => setIsUnLocked(false)}>
      <h1>Click to unlock</h1>
    </PhoneLockedScreen>
  );

  const renderUnlockedScreen = (
    <>
      <PhoneDisplay text={displayText} />
      <PredictionSection
        predictions={predictionWords}
        predictionsLoading={predictionWordsLoading}
        onSelectPrediction={onPredictionSelect}
      />
      <Keyboard
        onKeyPressed={onKeyPressed}
        onDeleteChar={onDeleteChar}
        show={!isUnLocked}
      />
    </>
  );

  return (
    <PhoneLayout>
      <PhoneContent>
        {isUnLocked ? renderLockedScreen : renderUnlockedScreen}
      </PhoneContent>
    </PhoneLayout>
  );
};

export default Phone;
