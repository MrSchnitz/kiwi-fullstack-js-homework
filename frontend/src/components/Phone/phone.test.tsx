import React from "react";
import { shallow, ShallowWrapper } from "enzyme";
import Phone, { PhoneProps } from "./phone.component";

describe("Phone component", () => {
  let wrapper: ShallowWrapper;
  let mockOnDeleteChar: any;
  let mockOnKeyPressed: any;
  let mockOnPredictionSelect: any;

  beforeEach(() => {
    mockOnDeleteChar = jest.fn();
    mockOnKeyPressed = jest.fn();
    mockOnPredictionSelect = jest.fn();

    const mockProps: PhoneProps = {
      onDeleteChar: mockOnDeleteChar,
      predictionWordsLoading: false,
      predictionWords: [],
      displayText: "this is good",
      onKeyPressed: mockOnKeyPressed,
      onPredictionSelect: mockOnPredictionSelect,
    };

    wrapper = shallow(<Phone {...mockProps} />);
  });

  it("should render Phone component", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should be unlocked when component is clicked", () => {
    wrapper.find("PhoneLockedScreen").simulate("click");

    expect(wrapper.find("PhoneContent"));
  });
});
