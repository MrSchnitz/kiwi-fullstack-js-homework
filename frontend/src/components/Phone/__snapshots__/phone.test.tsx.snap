// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Phone component should render Phone component 1`] = `ShallowWrapper {}`;
